lib/setupmanagers/qemuhck/qemu_machine.rb

```diff
@@ -161,33 +163,30 @@ def close
     STATES_JSON = 'lib/setupmanagers/qemuhck/states.json'
 
     def initialize(options)
+      # Sorbet requires to define all instance variables types in the constructor
+      @device_infos = T.let([], T::Array[Models::QemuHCKDevice])
+      @config_commands = T.let([], T::Array[String])
+      @pre_start_commands = T.let([], T::Array[String])
+      @post_stop_commands = T.let([], T::Array[String])
+      @define_variables = T.let({}, T::Hash[String, String])
       define_local_variables
 
       load_options(options)
       init_config
       apply_states
-      init_ports
-
-      @nm = NetworkManager.new(@id, @client_id, @machine, @logger)
-      @sm = StorageManager.new(@id, @client_id, @config, @options, @logger)
```

Comment:
We shouldn't move these variables away from `initialize`.

Instead, I suggest creating methods to derive `config_commands`, `pre_start_commands`, `post_stop_commands` and `define_variables` from `@device_infos` to eliminate the corresponding variables.

Reply:
Currently, `define_variables` is used in states, so we can add the `device_define_variables` method but the `define_variables` variable still must exist for `apply_state`. `config_commands`, `pre_start_commands`, `post_stop_commands` looks more device-specific.

Reply:
Here is my next suggestion:
* Merge `apply_states` into `init_config`. This will also eliminate `@states_config` as a bonus.
* Let `load_devices` return `T::Array[Model::QemuHCKDevice]` and assign it into `@device_infos`. This removes the mutation of `@device_infos`, an instance variable, and makes `QemuMachine` less stateful.

`QemuMachine` is too stateful so let's keep thinking about how to remove states.